pkgdown/build.R

```diff
@@ -0,0 +1,111 @@
+  system2("git", c("-C", temp_repo, "switch", "--detach", "@"))
+  build_version <- build_version_factory(temp_repo, versions, url, destination)
+
+  # NOTE: building the ref in root first, so pkgdown doesn't complain about a non-empty destination directory
+  root_index <- purrr::detect_index(versions, \(x) isTRUE(x$href == "/"))
+  purrr::walk(c(versions[root_index], versions[-root_index]), build_version)
+}
+
+validate_versions <- function(versions) {
+  expected_names <- c("ref", "href", "label")
+  n_root <- 0
+  purrr::walk(versions, function(version) {
+    diff <- setdiff(expected_names, names(version))
+    if (length(diff) > 0) {
+      stop("A version is missing the following keys: ", paste(diff, collapse = ", "))
+    }
+    if (isTRUE(version$href == "/")) {
+      n_root <<- n_root + 1
+      if (n_root > 1) {
+        stop("Exactly one version should have href set to '/'")
+      }
+    }
+  })
+}
+
+build_version_factory <- function(repo, versions, url, destination) {
+  navbar_template <- navbar_template_factory(versions, url)
+  destination <- fs::path_abs(destination)
+  extra_css_path <- fs::path_join(c(repo, "pkgdown", "extra.css"))
+
+  function(version) {
+    # Prepare a worktree for building
+    build_dir <- fs::file_temp("versioned-build-worktree-")
+    on.exit(system2("git", c("-C", repo, "worktree", "remove", "--force", build_dir))) # NOTE: --force because we add the navbar file
+    status <- system2("git", c("-C", repo, "worktree", "add", build_dir, version$ref))
+    if (status != 0) {
+      stop("Failed to create a worktree for ref ", version$ref)
+    }
+
+    # Write the navbar template and extra.css
+    template_dir <- fs::path_join(c(build_dir, "pkgdown", "templates"))
+    fs::dir_create(template_dir)
+    writeLines(navbar_template(version), fs::path_join(c(template_dir, "navbar.html")))
+    fs::file_copy(extra_css_path, fs::path_join(c(build_dir, "pkgdown", "extra.css")), overwrite = TRUE)
+
+    # NOTE: providing an absolute path to build_site won't work: https://github.com/r-lib/pkgdown/issues/2172
+    withr::with_dir(build_dir, {
+      pkgdown::build_site_github_pages(
+        override = list(
+          url = sub("/$", "", url_join(url, version$href)),
+          navbar = list(type = "light")
+        ),
+        dest_dir = fs::path_join(c(destination, version$href))
+      )
+    })
+  }
+}
+
+url_join <- function(url, path) {
+  paste(
+    sub("/$", "", url),
+    sub("^/", "", path),
+    sep = "/"
+  )
+}
+
+navbar_template_factory <- function(versions, url) {
+  navbar_code <- readLines("pkgdown/navbar_template.html")
+  index_current <- grep("___CURRENT_PLACEHOLDER___", navbar_code)
+  index_options <- grep("___OPTIONS_PLACEHOLDER___", navbar_code)
+  stopifnot(index_current < index_options)
+  wrap_label <- function(label) {
+    if (isTRUE(label)) {
+      label <- paste(desc::desc_get_version(), "(dev)")
```

Comment:
There are two subtly hidden assumptions here - `desc_get_version()` returns the correct version because:
1. It is called in the "parent repo" (outside of the worktree created for each version).
2. The "parent repo" is checked out at the current dev version.

I think it would be cleaner and more robust to explicitly add the dev version in the validation function (and rename it accordingly). We'd also avoid the special `label: true`.

Reply:
> add the dev version in the validation function
What do you mean by that?

I think we still need to have a way to use `desc_get_version()` to create the label. However, with fetching the latest patch version the labels might not be even needed. I opted for them to have everything explicitly listed. `label: true` was sort of an after-thought.

Reply:
Probably I wasn't quite clear here either. Yes, we want to call `desc_get_version()` to grab the development version. But I see a problem in the way we do it here:
1. It reads the version from the `DESCRIPTION` file in the **current working directory** which just happens to be the right place due to the way the code is structured (`navbar_template()` is called outside of `withr::with_dir()`).
2. Misleadingly, it isn't necessarily the version from `refs/remotes/origin/main` as defined in `versions.yml`; rather it is the version at the ref on which the `pkgdown.yml` workflow runs. Which, again, happens to be the right thing as we are going to run this workflow on pushes to `main`.

While (2) is quite unlikely to lead to any problems, I was actually bitten by (1) after refactoring the code a bit to attempt a `before_navbar` solution.

Anyway, I think it's fine to leave it as is. I think I'll send a PR with a `before_navbar` solution and tackle this along the way.